docs/docs/reference/extend/rules.md

```diff
@@ -0,0 +1,98 @@
+implemented encoding, which is not decided yet (see next section)
+
+## Syntax Changes
+
+The syntax of extensions is specified below as a delta with respect to the Scala syntax given [here](http://dotty.epfl.ch/docs/internals/syntax.html)
+
+New Keywords:
+
+    extension       common
+
+New Productions:
+
+    TmplDef           ::=  ...
+                        |  ‘extension’ ExtensionDef
+    ExtensionDef      ::=  id [ExtensionParams] 'for' Type ExtensionClause
+    ExtensionParams   ::=  [ClsTypeParamClause] [[nl] ImplicitParamClause]
+    ExtensionClause   ::=  [`:` Template]
+                        |  [nl] ‘{’ ‘def’ DefDef {semi ‘def’ DefDef} ‘}’
+
+    Modifier          ::=  ...
+                        |  ‘common’
+    Path              ::=  ...
+                        |  Path ‘.’ ‘common’
+
+
+## Type Checking Outline
+
+Here's an outline of the type checking rules and the auxiliary definitions implicitly added to programs. To keep things simple, this section only talks about the first-order case.
+The section on "Parameterized Typeclass Traits" talks a bit how to extend this to higher-kinds, and the[typeclass encoding test case](https://github.com/dotty-staging/dotty/blob/add-common/tests/pos/typeclass-encoding3.scala) provides further examples.
+
+### TypeClass Traits
+
+There is a new marker trait `scala.TypeClass`. Traits and classes inheriting directly or indirectly from `TypeClass` are called _typeclass traits_. They differ as follows from normal traits:
+
+ - They can have `common` definitions
+ - They can refer to the type `This`
+
+The companion object of a typeclass trait `TC` contains two compiler-generated members
+
+ - A type `Impl[T]` mapping an implementation type `T` of the trait to the type of the
+   corresponding implementation object, which coincides with `T`'s `common` part.
+
+ - A method `impl[T]` returning the implementation object of `T` for `TC`.
+
+### The `This` Type
+
+The `This` type is modeled as an abstract type. It is as if each typeclass trait
+was augmented with the following definition:
+
+    common type This
+
+Typeclass traits have an implicit conversion in scope that converts `This` to an
+instance of the typeclass trait.
+
+### Rules for Common
+
+Common definitions cannot see instance members of their enclosing trait. Common definitions in classes (but not traits) _can_ see the type parameters of the enclosing class.
+
+A Typeclass trait contains an instance member `common` that refers to the object containing the `common` members of the trait. If `p` refers to a typeclass trait, then `p.common` refers to this object.
+
+### Typeclass Implementation
+
+An implementation of a typeclass trait is a class or object that extends the trait (possibly that implementation is generated from an extension clause).
+An implementation `C` fixes the type of `This` to
+
+    type This = C
+
+unless `C` extends another implementation class `B` (in this case `B` has already defined `This`.)
```

Comment:
Back to the example above with `Option[A]` extending typeclass `Filter[A]`, if we want `Option` to implement typeclasses of different kinds, like `Filter` (or `Monad`) and `Semigroup`, we can't use the OOP typeclass-implementation style. Rust avoids this problem because it doesn't have higher kinds.